(GH-83) Turn off download progress

jrdnr opened the pull request:
Fixes Issue #83 "Turn off download progress"
Choco Versions >10.4 overwhelm log with download Progress data

Added Check for Choco version and add "--no-progress" to the Params list
 - Renamed Params variable for consistency across functions
 - Updated Choco comand execution formatting to be consistent across functions

jrdnr committed:
Choco Versions >10.4 overwhelm log with download Progress data

Added Check for Choco version and add "--no-progress" to the Params list
 - Renamed Params variable for consistency across functions
 - Updated Choco comand execution formatting to be consistent across functions

jrdnr committed:
Fix Pester Test run Failure

Adds "[Diagnostics.CodeAnalysis.SuppressMessage('PSAvoidUsingInvokeExpression','')]" to function UninstallPackage, to match other functions that use InvokeExpression to call Chco

ferventcoder commented on DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1:
```diff
@@ -262,17 +269,22 @@ function UninstallPackage
+        $chocoParams += " --version=`"$pVersion`""
     }
+    # Check if Chocolatey version is Greater than 0.10.4, and add --no-progress 
+    $chocoVer = [system.version](Get-ChocoInstalledPackage | Where-Object { $_.name -eq 'chocolatey' }).Version
```
Just run `choco -v` and inspect the output. Do keep in mind you can't do a System.Version on prereleases, so this would fail.

jrdnr replied:
Can you explain why [System.Version] would not work on prereleases?  I thought this was just a .net type fore version numbers that allows you to compare software versions with the assumption that there is a Major.Minor.Build.Revision numbering schema. 

If you set `$ChocoVer = '0.10.4'` it will auto cast it to type Version, I just wanted to be explicit to make sure that it did not end up as a string, as powershell would just be comparing string length if that happened.

ferventcoder replied:
The version would be `0.10.9-beta1` - cast that to a System.Version and check out the results. However, that Semantic Version is correct, it just doesn't cast to System.Version. You can split the number by the `-` and use the left side (index 0) for the cast into System.Version. That would work.

ferventcoder commented on DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1:
```diff
@@ -379,17 +392,23 @@ Function Upgrade-Package {
     }
-    $cmd = "choco upgrade $pName $chocoupgradeparams"
+    # Check if Chocolatey version is Greater than 0.10.4, and add --no-progress 
+    $chocoVer = [system.version](Get-ChocoInstalledPackage | Where-Object { $_.name -eq 'chocolatey' }).Version
```
Same comments as above.

ferventcoder reviewed (changes requested):
Pretty good, but finding the version of Chocolatey is quite a bit simpler, plus there may be times where (depending on how it was installed) there may not be a Chocolatey nupkg in the lib folder.

jrdnr commented:
The reason I used Get-ChocoInstalledPackage instead of choco -v, is because of the work done with caching the package list, running choco -v 15 times takes about 10.3 seconds, where as running Get-ChocoInstalledPackage takes about 1.6.  Pushing that up to 45 repetitions and choco -v is around 30 seconds and Get-ChocoInstalledPackage is still only at 1.8 seconds.

I'm sorry I don't know enough about how Choco works to follow what would happen if there were not a "Chocolatey nupkg in the lib folder" would this mean that `choco list -lo -r` would not return a value for chocolatey?

jrdnr committed:
Powershell cannot compair 0.10.4 to a version number

Powershell needs versions to be specified as strings in order to compair them to version numbers.
Updated to '0.10.4' insure accurate logic.

ferventcoder commented:
> I'm sorry I don't know enough about how Choco works to follow what would happen if there were not a "Chocolatey nupkg in the lib folder" would this mean that choco list -lo -r would not return a value for chocolatey?

It depends on how Chocolatey was installed, but folks are free to install it in their own ways and while we wish everything was a good steward about getting the nupkg in the proper place, that isn't always true. Above that, for the longest time even our scripts were not doing it correctly, so some of the configuration managers out there were also implementing it wrong. So we can't assume that the nupkg is there.


> The reason I used Get-ChocoInstalledPackage instead of choco -v, is because of the work done with caching the package list, running choco -v 15 times takes about 10.3 seconds, where as running Get-ChocoInstalledPackage takes about 1.6. Pushing that up to 45 repetitions and choco -v is around 30 seconds and Get-ChocoInstalledPackage is still only at 1.8 seconds.

This is a great point. Perhaps cache the version output as well? I think it will be even faster than iterating the cache of installed packages and it will also be more accurate due to the above constraints.

Thoughts?

ferventcoder commented on DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1:
```diff
@@ -262,17 +269,22 @@ function UninstallPackage
     }
+    # Check if Chocolatey version is Greater than 0.10.4, and add --no-progress 
+    $chocoVer = [system.version](Get-ChocoInstalledPackage | Where-Object { $_.name -eq 'chocolatey' }).Version
+    if ($chocoVer -ge '0.10.4') {
```
For explicit sake, I'd like this to be `if ($chocoVer -ge ([System.Version]'0.10.4')) {` ~~, otherwise you may be comparing two string values. It looks correct and works most of the time, but then comes very hard to debug issues and everything looks correct~~.

Actually, it looks like it converts it to System.Version to make the check instead of converting `$chocoVer` to String format for the operation.

~~~
[System.Version]$chocoVersion = '0.10.010'
$chocoVersion.ToString() -ge '0.10.09'
$chocoVersion.ToString() -ge '0.10.011'
$chocoVersion -ge ([System.Version]'0.10.011')
~~~

Good on PowerShell there. It may not be a bad idea to be explicit anyway in case that changes or is different in any future version of PowerShell.

jrdnr replied:
I had verified that it was comparing [System.Version] types but I would agree being explicit is better than relying on magic.

jrdnr committed:
Implemented caching choco version
Ensured explicitly casting '0.10.4' to Version

ferventcoder commented:
> Is there a reasonable way to know if choco gets upgraded or should I just go with the timer?

I think the timer is going to be fine.

ferventcoder commented on DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1:
```diff
@@ -426,4 +442,26 @@ function Get-ChocoInstalledPackage ($action) {
+        if ( $cacheSec -lt (Get-Item $chocoVersion -ErrorAction SilentlyContinue).LastWriteTime ) {
+                $res = Import-Clixml $chocoVersion
+        } else {
+            $res = [version](choco -v) | Export-Clixml -Path $chocoVersion
```
That switch to System.Version won't work on `0.10.9-beta1`. Might be best to do that split by `-` and choosing the left side prior to conversion.

jrdnr replied:
Oh, perfect, that's the detail I was missing, I didn't know how the choco beta versions were formatted.

jrdnr committed:
Choco appends '-beta' to beta version numbers
Added parsing to split on '-' and only use the first section.

ferventcoder commented on DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1:
```diff
@@ -226,23 +226,28 @@ function InstallPackage
+        $chocoParams += " $cParams"
+    }
+    # Check if Chocolatey version is Greater than 0.10.4, and add --no-progress 
+    if ((Get-ChocoVersion) -ge [version]('0.10.4')){
```
Not to be too difficult, but fully qualify this. I've run into issues where things were not fully qualified causing issues in different versions of PowerShell. `[System.Version]` 

Thanks!

jrdnr replied:
will do

ferventcoder commented on DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1:
```diff
@@ -426,4 +442,28 @@ function Get-ChocoInstalledPackage ($action) {
+            $res = Import-Clixml $chocoVersion
+        } else {
+            $cmd = choco -v
+            $res = [version]($cmd.Split('-')[0])
```
Note that your commit message is incorrect. It's not `-beta`, it's really anything that comes with `-` attached to it. It's the concept of prereleases, which has been around since NuGet was first released and subsequently when Chocolatey was first released.

ferventcoder replied:
It could be `-alpha.20171212`, or really any value.

jrdnr committed:
Updated [Version] to Fully Qualified [System.Version]

jrdnr commented:
@ferventcoder Let me know if there are any other items that need to be addressed before this can be merged.  If this is good to go as implemented I can squash it down and correct the commit messages that are inaccurate.

pauby commented:
@ferventcoder Can you take a look at this and see if there is anything else needed since your review?

pauby commented on DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1:
```diff
@@ -426,4 +442,28 @@ function Get-ChocoInstalledPackage ($action) {
     Return $res
 }
 
+function Get-ChocoVersion ($action) {
```
Can we use advanced functions here in keeping with the majority of the code?

jrdnr replied:
New Functions updated to match the style of "Advanced Function" in use in the rest of the file.  Let me know if there are any other tweaks needed

pauby reviewed (commented):
@jrdnr I've gone through and resolved the conversations by @ferventcoder - just one small point from me, where we can use Advanced Functions. After that I'm happy to give this a spin and merge.

jrdnr committed:
Converted simple functions to "Advanced Functions" to match the rest of code

pauby commented on DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1:
```diff
@@ -417,42 +417,54 @@ Function Upgrade-Package {
-        Remove-Item $ChocoInstallList -Force
-        $res = $true
-    } else {
+    if ( -not $Purge) {
```
The previous code if / else was reversed in these changes. Can we have the `if ($Purge.IsPresent)` at the top so it matches the previous code?

pauby commented on DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1:
```diff
@@ -417,42 +417,54 @@ Function Upgrade-Package {
-        Remove-Item $chocoVersion -Force
-        $res = $true
-    } else {
+    if ( -not $Purge) {
```
Same here?

pauby replied:
Thanks for reversing that - I think you missed removing the `-not` from this?

jrdnr replied:
gah.. I saw that right after I pushed the commit and fixed it but forgot to re-commit it.  proper fix coming in a second.

jrdnr replied:
Is testing $Purge.IsPresent, considered the correct way for test vs just `if ($Purge)`?  I was intending to update that but accidentally committed before being done making the requested updates.

pauby replied:
I think it's a personal taste because both work. For me, what I see `$Purge.IsPresent` I know `$Purge` is a switch. If I just see `$Purge` on it's own it could be anything.

jrdnr committed:
Revered if Purge.IsPresent checks back to origional order

jrdnr committed:
Corrected reversed logic in `if (-not $Purge)
Using `$Purge.IsPresent` to be more explicit.
